api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,616 @@
+
+		for _, consumer := range resourceConsumers {
+			hasRoom := false
+
+			for _, provider := range providerDistribution.Providers {
+				if provider.HasResourcesFor(consumer) {
+					hasRoom = true
+					if err := provider.SubtractResourcesFor(consumer); err != nil {
+						return nil, []error{err}
+					}
+					break
+				}
+			}
+
+			if !hasRoom {
+				newProvider, err := e.calculateNewProvider(clusterName)
+				if err != nil {
+					return nil, []error{err}
+				}
+
+				newProvider.AgentIsConnected = true
+				newProvider.Status = ecs.ContainerInstanceStatusActive
+
+				if !newProvider.HasResourcesFor(consumer) {
+					text := fmt.Sprintf("Resource '%s' cannot fit into an empty provider!", consumer.ID)
+					text += "\nThe instance size in your environment is too small to run this resource."
+					text += "\n Please increase the instance size for your environment or remove this resource."
+					err := errors.Newf(errors.IncompatibleConsumerAndProvider, text)
+					providerDistribution.Errs = append(providerDistribution.Errs, err)
+					continue
+				}
+
+				if err := newProvider.SubtractResourcesFor(consumer); err != nil {
+					return nil, []error{err}
+				}
+
+				providerDistribution.Providers = append(providerDistribution.Providers, newProvider)
+			}
+		}
+	}
+
+	p := findOptimalProviderDistribution(providerDistributions)
+	return p.Providers, p.Errs
+}
+
+func findOptimalProviderDistribution(providerDistributions []ProviderDistribution) ProviderDistribution {
+	if len(providerDistributions) == 1 {
+		return providerDistributions[0]
+	}
+
+	shortestLength := len(providerDistributions[0].Providers)
+	shortestDistribution := providerDistributions[0]
+
+	for _, p := range providerDistributions[1:] {
+		l := len(p.Providers)
+		if l < shortestLength {
+			shortestLength = l
+			shortestDistribution = p
+		}
+	}
+
+	return shortestDistribution
+}
+
+func (e *EnvironmentScaler) getContainerResourceFromDeploy(deployID string) ([]scaler.ResourceConsumer, error) {
+	if consumers, ok := e.deployCache.Getf(deployID); ok {
+		return consumers.([]scaler.ResourceConsumer), nil
+	}
+
+	taskDefinition, err := e.getTaskDefinitionFromDeployID(deployID)
+	if err != nil {
+		return nil, err
+	}
+
+	consumers := make([]scaler.ResourceConsumer, len(taskDefinition.ContainerDefinitions))
+	for i, taskDefinition := range taskDefinition.ContainerDefinitions {
+		var cpu int
+		var memory bytesize.Bytesize
+
+		if c := int(aws.Int64Value(taskDefinition.Cpu)); c != 0 {
+			cpu = c
+		}
+
+		if m := aws.Int64Value(taskDefinition.MemoryReservation); m != 0 {
+			memory = bytesize.MiB * bytesize.Bytesize(m)
+		}
+
+		if m := aws.Int64Value(taskDefinition.Memory); m != 0 {
+			memory = bytesize.MiB * bytesize.Bytesize(m)
+		}
+
+		ports := []int{}
+		for _, p := range taskDefinition.PortMappings {
+			if hostPort := int(aws.Int64Value(p.HostPort)); hostPort != 0 {
+				ports = append(ports, hostPort)
+			}
+		}
+
+		id := "NewConsumerFromDeploy::" + deployID
```

Comment:
`NewConsumerFromDeploy` isn't helpful, and we should have the container name as well. 
```
id := fmt.Sprintf("Deploy: %s, Container: %s", deployID, conatiner.Name)
```

We used to have the deploy copy # as well